src/Workspaces/CSharp/Portable/Extensions/ContextQuery/CSharpSyntaxContext.cs

```diff
@@ -283,6 +283,27 @@ public bool IsTypeAttributeContext(CancellationToken cancellationToken)
             return false;
         }
 
+        public bool IsTypeAttributeContextBefore(SyntaxKind syntaxKind, CancellationToken cancellationToken)
+        {
+            // cases:
+            //    [ |
+            //    class C {
+            var token = TargetToken;
+
+            // Note that we pass the token.SpanStart to IsTypeDeclarationContext below. This is a bit subtle,
+            // but we want to be sure that the attribute itself (i.e. the open square bracket, '[') is in a
+            // type declaration context.
+            if (token.Kind() == SyntaxKind.OpenBracketToken &&
+                (token.GetNextToken().Kind() == syntaxKind || token.Parent.Kind() == SyntaxKind.AttributeList) &&
+                SyntaxTree.IsTypeDeclarationContext(
+                    token.SpanStart, contextOpt: null, validModifiers: null, validTypeDeclarations: SyntaxKindSet.ClassStructTypeDeclarations, canBePartial: false, cancellationToken: cancellationToken))
```

Comment:
so, in general, these types of exceptionally complex expression are just hard to read and maintain.  Specifically because of the mixing and matching of the  &&s and ||s.

Reply:
So, i think a cleaner way to do this would be to have the following:

1. scrap this method.
2. Go to AssemblyKeywordRecommender and just change IsValidContext to be:

```c#
            var token = this.TargetToken;

            // Note that we pass the token.SpanStart to IsTypeDeclarationContext below. This is a bit subtle,
            // but we want to be sure that the attribute itself (i.e. the open square bracket, '[') is in a
            // type declaration context.
            if (token.Kind() != SyntaxKind.OpenBracketToken ||
                token.Parent.Kind() != SyntaxKind.AttributeList)
            {
               return false;
            }

            var attributeList= token.Parent;
            var parent = attributeList.Parent;
            return parent is CompilationUnit || parent is NamespaceDeclaration || parent is TypeDeclarationSyntax;
        }
```

Reply:
Basically, the bug is not in IsTypeAttributeContext.  The bug is that the AssemblyKeywordRecommender is using that helper at all.

Note: i would also place the above code in its own helper.  You'll need to use that in the ModuleKeywordRecommender as well.

Let me know if that helps!